sway/input/keyboard.c

```diff
@@ -494,6 +513,49 @@ static void handle_keyboard_group_key(struct wl_listener *listener,
 	handle_key_event(sway_group->seat_device->keyboard, data);
 }
 
+static void handle_keyboard_group_enter(struct wl_listener *listener,
+		void *data) {
+	struct sway_keyboard_group *sway_group =
+		wl_container_of(listener, sway_group, enter);
+	struct sway_keyboard *keyboard = sway_group->seat_device->keyboard;
+	struct wl_array *keycodes = data;
+
+	uint32_t *keycode;
+	wl_array_for_each(keycode, keycodes) {
+		struct key_info keyinfo;
+		update_keyboard_state(keyboard, *keycode, WLR_KEY_PRESSED, &keyinfo);
+	}
+}
+
+static void handle_keyboard_group_leave(struct wl_listener *listener,
+		void *data) {
+	struct sway_keyboard_group *sway_group =
+		wl_container_of(listener, sway_group, leave);
+	struct sway_keyboard *keyboard = sway_group->seat_device->keyboard;
+	struct wl_array *keycodes = data;
+
+	bool pressed_sent = false;
+
+	uint32_t *keycode;
+	wl_array_for_each(keycode, keycodes) {
+		struct key_info keyinfo;
+		update_keyboard_state(keyboard, *keycode, WLR_KEY_RELEASED, &keyinfo);
+
+		pressed_sent |= update_shortcut_state(&keyboard->state_pressed_sent,
+				*keycode, WLR_KEY_RELEASED, keyinfo.keycode, 0);
+	}
+
+	if (!pressed_sent) {
+		return;
+	}
+
+	// Refocus the focused node so the keyboard state gets synced
```

Comment:
Does this need to consider unmanaged Xwayland surfaces as well? I recall from [here](https://github.com/Xyene/sway/blob/a7bcfc56021c192975c7a5ae5f22d1d0770e37ea/sway/desktop/xwayland.c#L102-L118) that the focus for these are handled differently since they don't have `sway_node`s associated with them.

Reply:
I pushed a new version which handles when the currently focused surface is a layer surface or unmanaged surface.